File: lr/lr.r
# Multi-class Logistic Regression + Stochastic Gradient Descent for R
# (c)2011 Nakatani Shuyo / Cybozu Labs, Inc.
# This code is available under the MIT Licence.

# コマンドライン処理
commandline <- commandArgs(TRUE)
chart <- "--chart" %in% commandline
i <- match("-i", commandline)
if (is.na(i)) {
	I <- 1
} else {
	I <- as.numeric(commandline[i + 1])
}

# iris dataset
xlist <- scale(iris[1:4])
tlist <- cbind(
	ifelse(iris[5]=="setosa",1,0),
	ifelse(iris[5]=="versicolor",1,0),
	ifelse(iris[5]=="virginica",1,0)
)
N <- nrow(xlist) # データ件数

# 事後確率
y <- function(phi, w) {
	y <- c(phi %*% w)
	y <- exp(y - max(y))  # exp の中身から、その最大値を引く(オーバーフロー対策)
	return(y / sum(y))
}

# 誤差関数＆勾配
En <- function(phi, t, w) -log(sum(y(phi, w) * t))
dEn <- function(phi, t, w) outer(phi, y(phi, w) - t)

inference <- function(title, xlist, tlist, phi) {
	PHI <- t(apply(xlist, 1, phi))  # NxM - design matrix
	M <- ncol(PHI)  # 特徴数(特徴空間の次元)
	K <- ncol(tlist) # クラス数

	for (i in 1:I) {
		# 重み初期化
		w <- matrix(rnorm(M * K), M)

		eta <- 0.1  # 学習率
		while (eta > 0.0001) {
			for(n in sample(N)) {
				w <- w - eta * dEn(PHI[n,], tlist[n,], w)  # 確率的勾配降下法
			}
			eta <- eta * 0.95
		}

		ylist <- t(apply(PHI, 1, function(phi) y(phi, w)))
		error <- sum(sapply(1:nrow(PHI), function(n) En(PHI[n,], tlist[n,], w)))
		cat(sprintf("%s: error=%.3f", title, error), "\n")

		# 可視化
		if (chart) {
			pairs(xlist, col=rgb(ylist), main=title)
			plot(xlist[,c(1,2)],
				col=rgb(ylist),
				pch=(tlist %*% c(17,16,22)),
				main=title,
				sub=sprintf("Negative Log Likelihood = %.3f", error)
			)
		}
	}

	return(w)
}

if (chart) png(width=640, height=640)

# 線形特徴関数
phi <- function(x) c(1, x[1], x[2], x[3], x[4])
w <- inference("Linear Features", xlist, tlist, phi)

# 二次特徴関数
phi <- function(x) c(1, x[1], x[2], x[3], x[4],
	x[1]*x[1], x[1]*x[2], x[1]*x[3], x[1]*x[4], x[2]*x[2],
	x[2]*x[3], x[2]*x[4], x[3]*x[3], x[3]*x[4], x[4]*x[4])
w <- inference("Quadratic Features", xlist, tlist, phi)

# RBF 特徴関数
for (s in 1:10) {
	phi <- function(x) {
		c <- seq(-2.5,2.5,by=1)
		d <- outer(c,x,"-")^2
		return(exp(-c(0, outer(c(outer(c(outer(d[,1],d[,2],"+")),d[,3],"+")),d[,4],"+"))/s))
	}
	w <- inference(sprintf("RBF Features (s=%d)", s), xlist, tlist, phi)
}

if (chart) dev.off()

File: lr/lr.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Logistic Regression + Steepest Descent + FOBOS L1
# This code is available under the MIT License.
# (c)2011 Nakatani Shuyo / Cybozu Labs Inc.

import numpy

def linear_features():
    return lambda x: numpy.concatenate(([1], x))

class LR:
    def __init__(self, xlist, tlist, feature_func):
        self.tlist = tlist
        self.N, self.K = tlist.shape
        self.PHI = numpy.apply_along_axis(feature_func, 1, xlist)
        self.M = self.PHI.shape[1]
        self.w = numpy.random.randn(self.K, self.M)

    def predict(self, phi):
        y = numpy.dot(self.w, phi)
        y = numpy.exp(y - y.max())
        return y / y.sum()

    def error(self, phi, t):
        return -numpy.log(numpy.inner(self.predict(phi), t))

    def error_full(self):
        return numpy.sum([self.error(phi, self.tlist[n,]) for n, phi in enumerate(self.PHI)])

    def error_gradient(self, phi, t):
        return numpy.outer(self.predict(phi) - t, phi)

    def error_gradient_full(self):
        return numpy.sum([self.error_gradient(phi, self.tlist[n,]) for n, phi in enumerate(self.PHI)], axis=0)

    def inference_SGD(self, eta, reg=0, reg_coef=0.1):
        """SGD iteration once"""
        for n in numpy.random.shuffle(numpy.arange(self.N)):
            phi = self.PHI[n,]
            self.w -= eta * self.error_gradient(phi, self.tlist[n,])
            if reg == 1:
                self.fobos_l1(eta * reg_coef)
            elif reg == 2:
                self.fobos_l2(eta * reg_coef)

    def inference_SD(self, eta, reg=0, reg_coef=0.1):
        """Steepest Descent iteration once"""
        self.w -= eta * self.error_gradient_full()
        if reg == 1:
            self.fobos_l1(eta * reg_coef)
        elif reg == 2:
            self.fobos_l2(eta * reg_coef)

    def fobos_l1(self, reg_coef):
        self.w = (self.w - reg_coef) * (self.w > reg_coef) + (self.w + reg_coef) * (self.w < -reg_coef)

    def fobos_l2(self, reg_coef):
        self.w /= 1.0 + reg_coef

def main():
    import optparse
    parser = optparse.OptionParser()
    #parser.add_option("-f", dest="filename", help="corpus filename")
    parser.add_option("-i", dest="iteration", type="int", help="iteration count", default=10)
    parser.add_option("-r", dest="reg", type="int", help="regularization(r=0,1,2)", default=0)
    parser.add_option("-l", dest="reg_coef", type="float", help="regularization coefficient", default=0.1)
    parser.add_option("--sgd", dest="sgd", help="SGD inference", action="store_true", default=False)
    parser.add_option("--seed", dest="seed", type="int", help="random seed")
    (options, args) = parser.parse_args()

    #if not options.filename: parser.error("need corpus filename(-f)")
    #corpus = vocabulary.load_file(options.filename)
    if options.seed != None:
        numpy.random.seed(options.seed)

    iris_data = numpy.loadtxt('../data/iris.data.txt', delimiter=",", usecols=(0,1,2,3))
    iris_label = numpy.loadtxt('../data/iris.data.txt', delimiter=",", usecols=(4,), dtype='S16')

    # normalization
    xlist = iris_data - numpy.mean(iris_data, axis=0)
    xlist /= numpy.std(xlist, axis=0)

    # target as 1 of K representation
    labels = {'Iris-setosa':[1,0,0], 'Iris-versicolor':[0,1,0], 'Iris-virginica':[0,0,1]}
    tlist = numpy.array([labels[x] for x in iris_label])

    lr = LR(xlist, tlist, linear_features())

    eta = 0.1
    reg_coef = options.reg_coef
    for i in range(options.iteration):
        if options.sgd:
            lr.inference_SGD(eta, options.reg, reg_coef)
        else:
            lr.inference_SD(eta, options.reg, reg_coef)

        print lr.error_full()
    print lr.w

if __name__ == "__main__":
    main()
